### PR TITLE
S1t display

### PR DESCRIPTION
**Pull request by Zedonk1995**
The webpage is not responsive yet, that job is my next task which I'll make a separate branch for.  All my functions have been unit tested.

### CODE REVIEW

File: index.php [hunk 1]
```diff
@@ -1,3 +1,53 @@
...
+            </header>
+        </h1>
+
+        <main>
+            <section>
+                <h2>
+                    Collection Contents
+                </h2>
+
+                <list>
```
Comment (MattW-R) [inline index.php:32]: Is list a real tag? Probably wanted ul?

File: functions.php [hunk 2]
```diff
@@ -12,14 +9,121 @@
...
+ * @return string returns the bullet-pointed list of data names.
+ */
+function displayWeaponNames(array $data) : string
+{
+    $result = '';
+    foreach($data as $datum)
+    {
+            if(validateWeaponName($datum))
+            {
+                $result .= displayWeaponName($datum);
```
Comment (MattW-R) [inline functions.php:78]: Better to not call call your functions within functions - For unit testing & single functionality reasons

File: functions.php [hunk 3]
```diff
@@ -12,14 +9,121 @@
...
+ * @return string returns a string that contains all of the weapons and their stats
+ */
+function displayWeaponData(array $weaponData) : string
+{
+    $result = '';
+    if(validateWeaponData($weaponData))
+    {
+        foreach($weaponData as $weapon)
+        {
+            $result .= displayWeaponDatum($weapon);
```
Comment (MattW-R) [inline functions.php:117]: Same as above. Better to not call your functions within function.

File: functions.php [hunk 4]
```diff
@@ -12,14 +9,121 @@
...
+ *
+ * @param array $data the data that will get returned.  This data must have a 'name' key.
+ * @return string returns the bullet-pointed list of data names.
+ */
+function displayWeaponNames(array $data) : string
+{
+    $result = '';
+    foreach($data as $datum)
+    {
+            if(validateWeaponName($datum))
```
Comment (MattW-R) [inline functions.php:76]: Probably don't need to create entire function for this. Just use array_key_exists as above in the function declaration & add a comment to say what you are doing & then you are not calling your own functions within functions.

File: tests/functions.php [hunk 5]
```diff
@@ -1,3 +1,168 @@
...
+class Functions extends TestCase
+{
+    public function testSuccessValidateWeaponName()
+    {
+        $expected = true;
+        $input = ['name' => 'Firestorm', 'damage' => 10];
+        $case = validateWeaponName($input);
+        $this->assertEquals($expected,$case);
+    }
+
```
Comment (MattW-R) [inline tests/functions.php:15]: Could add failure test here with $input = [];

File: tests/functions.php [hunk 6]
```diff
@@ -1,3 +1,168 @@
...
+            'rate_of_fire' => '5',
+            'ammo_consumption' => '4',
+            'ammo_efficiency' =>  '2',
+            'default_weapon_number' => '3',
+            'image_url' => 'some kind of url'
+            ]];
+        $case = validateWeaponData($input);
+        $this->assertEquals($expected,$case);
+    }
+    public function testSuccess2ValidateWeaponData()
```
Comment (MattW-R) [inline tests/functions.php:38]: This should probably be named testFailureValidateWeaponData(). Could add a second example/assertion using an input array with missing keys for extra rigour.

File: tests/functions.php [hunk 7]
```diff
@@ -1,3 +1,168 @@
...
+    }
+
+    public function testSuccessDisplayWeaponName()
+    {
+        $expected = '<li>Firestorm</li>';
+        $input = ['name' => 'Firestorm', 'damage' => 10];
+        $case = DisplayWeaponName($input);
+        $this->assertEquals($expected,$case);
+    }
+
```
Comment (MattW-R) [inline tests/functions.php:68]: Could add failure test using $input = [];

File: tests/functions.php [hunk 8]
```diff
@@ -1,3 +1,168 @@
...
+            'rate_of_fire' => '5',
+            'ammo_consumption' => '4',
+            'ammo_efficiency' =>  '2',
+            'default_weapon_number' => '3',
+            'image_url' => 'some kind of url'
+        ];
+        $case = displayWeaponDatum($input);
+        $this->assertEquals($expected,$case);
+    }
+
```
Comment (MattW-R) [inline tests/functions.php:114]: Could add failure test with $input = [];

File: index.php [hunk 9]
```diff
@@ -1,3 +1,53 @@
...
+            My Collection of Old School FPS weapons
+        </title>
+    </head>
+
+    <body>
+        <h1>
+            <header>
+                Collection of Old School FPS weapons
+            </header>
+        </h1>
```
Comment (MattW-R) [inline index.php:24]: h1 should be inside header

File: tests/functions.php (hunk 5 above)
Comment (Zedonk1995) [reply tests/functions.php:15]: I don't think this is a failure test.  I'd consider any input array as a proper use of this function

File: index.php (hunk 9 above)
Comment (Zedonk1995) [reply index.php:24]: done

File: index.php (hunk 1 above)
Comment (Zedonk1995) [reply index.php:32]: yh, i put in list by mistake

File: tests/functions.php (hunk 6 above)
Comment (Zedonk1995) [reply tests/functions.php:38]: see above

File: tests/functions.php (hunk 8 above)
Comment (Zedonk1995) [reply tests/functions.php:114]: see above

File: functions.php (hunk 2 above)
Comment (Zedonk1995) [reply functions.php:78]: I've asked Charlie and he's done a similar thing in his code so I think I'll leave it as is (so I can test the other functions)

File: tests/functions.php (hunk 6 above)
Comment (Zedonk1995) [reply tests/functions.php:38]: I don't think this is a failure test. I'd consider any input array as a proper use of this function

**Pull request merged**